From 7d6afad324aedb763bc5986fe123a18835db9c0e Mon Sep 17 00:00:00 2001
From: Tomas Popela <tpopela@redhat.com>
Date: Tue, 18 Oct 2016 13:16:21 +0200
Subject: [PATCH 455/595] Bug 772918 - Using keyboard shortcuts undo/redo is
 broken

If redoing an INPUT event that was done in the middle of the text we need to
move one character backward as the range is pointing after the character and
not before it - for INPUT events we don't save the before coordinates.
---
 .../web-extension/e-editor-undo-redo-manager.c     | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/modules/webkit-editor/web-extension/e-editor-undo-redo-manager.c b/modules/webkit-editor/web-extension/e-editor-undo-redo-manager.c
index 3e23716933..e60feb40fb 100644
--- a/modules/webkit-editor/web-extension/e-editor-undo-redo-manager.c
+++ b/modules/webkit-editor/web-extension/e-editor-undo-redo-manager.c
@@ -798,6 +798,28 @@ undo_delete (EEditorPage *editor_page,
 		else
 			range = get_range_for_point (document, event->after.start);
 
+		/* If redoing an INPUT event that was done in the middle of the
+		 * text we need to move one character backward as the range is
+		 * pointing after the character and not before it - for INPUT
+		 * events we don't save the before coordinates. */
+		if (event->type == HISTORY_INPUT) {
+			glong start_offset;
+			WebKitDOMNode *start_container;
+
+			start_offset = webkit_dom_range_get_start_offset (range, NULL);
+			start_container = webkit_dom_range_get_start_container (range, NULL);
+
+			if (WEBKIT_DOM_IS_CHARACTER_DATA (start_container) &&
+			    start_offset != webkit_dom_character_data_get_length (WEBKIT_DOM_CHARACTER_DATA (start_container))) {
+				webkit_dom_range_set_start (
+					range,
+					start_container,
+					start_offset > 0 ? start_offset - 1 : 0,
+					NULL);
+				webkit_dom_range_collapse (range, TRUE, NULL);
+			}
+		}
+
 		webkit_dom_range_surround_contents (range, WEBKIT_DOM_NODE (element), NULL);
 		webkit_dom_dom_selection_remove_all_ranges (dom_selection);
 		webkit_dom_dom_selection_add_range (dom_selection, range);
-- 
2.15.0

